### DISCUSSION TITLE
2023-08-07 정기 백엔드 회의

### DISCUSSION THREAD

**Discussion by scarf005** [jiphyeonjeon-42/backend#646] [General]
진행 시간: 2023-08-07 (월) 14:00
진행 방식: https://github.com/jiphyeonjeon-42/backend/discussions/645

안건을 자유롭게 추가해주세요.

**Comment by nyj001012** [comment]
1. stock, histories 진행 상황 공유
2. 앞으로의 계획..?

**Comment by scarf005** [comment]
# 07-18~08-07 [포스트모템](https://kukim.tistory.com/140)

- 07-18 (1주차): https://www.collabee.co/project/163329/issue/postDetail/2386529
- 07-24 (2주차): https://www.collabee.co/project/163329/issue/postDetail/2392904
- 07-31 (3주차): https://github.com/jiphyeonjeon-42/backend/discussions/613

'타입 안전한 백엔드'를 목표로 백엔드 V2 리팩터링 팀이 본격적으로 시작한지도 어느덧 3주가 되었습니다. 지금까지 좋았던 점과 아쉬운 점에 대해 이야기해보고, 원인 분석을 통해 앞으로 집현전 백엔드 팀이 나아갈 방향에 대해 이야기해보는 시간을 가져보면 좋을 것 같습니다.

**Comment by scarf005** [comment]
# 아쉬웠던 점

## 지나치게 많고 복잡한 목표

- [pnpm 모노레포](https://github.com/jiphyeonjeon-42/backend/pull/589) 전환
- typeorm find 메서드 사용
- ts-rest 사용
- zod 사용
- vite-node 사용
- ts-pattern 사용
- DI 사용
- 테스팅 (vitest)
- [v2 경로 추가, 절대경로 사용](https://github.com/jiphyeonjeon-42/backend/pull/584)
- v2 API 경로 재설계
- [프로젝트 보드](https://github.com/orgs/jiphyeonjeon-42/projects/8?query=is%3Aopen+sort%3Aupdated-desc) 사용
- 개발 서버 추가

거의 전부 제가 제안한 내용들인데, 욕심을 버리지 못하고 계획들을 추가해 작업을 시작하려면 위의 모든 것을 공부해야 했고, 때문에 입문이 어려워지고, 일정이 미뤄지게 된것 같습니다.

## 충분한 고민 없는 빠른 변화

- #584 
- #591
- #640

프로젝트 실행에 많은 영향을 주는 변화였음에도 충분한 소통 없이 바로 진행해 #648 와 같은 문제가 발생했습니다.

## 성공적이지 못한 DI로 인해 지나치게 복잡해진 기능 추가

### 컨트롤러

https://github.com/jiphyeonjeon-42/backend/blob/a23943abbe9d79489698e8db5d1c9d5a90014d70/backend/src/v2/reviews/controller/controller.ts#L66-L82

DI를 추가했음에도 원래 목표와는 달리 컨트롤러를 테스트하는데 큰 도움이 되지 않았습니다. 이는 이미 ts-rest가 제공하는 기능에 의해

- [ts-rest contracts](https://ts-rest.com/docs/core/)에서 자동으로 입력값 파싱과 반환값을 검사
- 컨트롤러가 하는 일이 `서비스 호출 -> 호출 결과에 따라 결과 반환`으로 간소화

컨트롤러의 유닛 테스트를 하는 장점이 줄었기 때문입니다. 반면 변경된 구조에서는 기능 추가를 위해서는

1. 컨트롤러 타입을 정의
2. [impl.ts](https://github.com/jiphyeonjeon-42/backend/blob/a23943abbe9d79489698e8db5d1c9d5a90014d70/backend/src/v2/reviews/controller/impl.ts#L9-L14) 파일을 추가
3. 실행하기 전 [초기화](https://github.com/jiphyeonjeon-42/backend/blob/a23943abbe9d79489698e8db5d1c9d5a90014d70/backend/src/v2/reviews/impl.ts#L17-L35)

를 해 주어야 하는 등 명시적인 장점 없이 코드 작성과 이해에 드는 노력이 크게 늘어났습니다.

### 서비스

마찬가지로, 과정이 크게 복잡해진 반면 서비스에 적용한 DI 또한 테스트에 큰 도움이 되지 못했습니다. 이는 

- [개발 서버에서](https://www.tonic.ai/blog/using-docker-to-manage-your-test-database) [통합 테스트 진행](https://www.tedinski.com/2019/03/19/testing-at-the-boundaries.html)이라는 기존 목표랑 상충하며
- 의존성이 typeorm Repository 타입이므로 유닛 테스트를 진행하려면 별도의 typeorm mocking 라이브러리를 추가해야 했기 때문입니다.

이렇게 새 디자인 패턴이 큰 도움이 되지 못한 반면 컨트롤러처럼 구현 과정은 복잡해졌습니다:

1. 서비스 타입을 정의

https://github.com/jiphyeonjeon-42/backend/blob/a23943abbe9d79489698e8db5d1c9d5a90014d70/backend/src/v2/reviews/service/index.ts#L5-L32

2. 서비스 구현을 정의

https://github.com/jiphyeonjeon-42/backend/blob/a23943abbe9d79489698e8db5d1c9d5a90014d70/backend/src/v2/reviews/service/service.ts#L12-L29

3. [impl.ts](https://github.com/jiphyeonjeon-42/backend/blob/a23943abbe9d79489698e8db5d1c9d5a90014d70/backend/src/v2/reviews/service/impl.ts) 정의
4. 최상위 [impl.ts](https://github.com/jiphyeonjeon-42/backend/blob/a23943abbe9d79489698e8db5d1c9d5a90014d70/backend/src/v2/reviews/impl.ts) 에서 다시 의존성 주입

**Comment by scarf005** [comment]
## 2주간 가능한 날

https://www.when2meet.com/?20801850-noV6Z

가능한지의 기준
- github나 슬랙 메시지 답장 가능
- 기능 개발에 참여 가능
- 코드 리뷰에 참여 가능

**Comment by scarf005** [comment]
# 정리

## 진행이 느렸던 이유

- 과제로 바쁨
- 장기 계획 부재
- 프로젝트 입문이 복잡함
- v2 수정이 v1 서버에 영향을 미칠수 있음

## 진행을 빠르게 하려면

- 시간 낼수 있는 날에는 알림을 잘 받으려고 노력하기!
- V2 API 스펙을 빠르게 정리하기
- v2에 변화가 있을때 서버가 서버 터지는 것을 막기 위해서 v2를 v1와 별개의 서버로 운영하기

**Comment by scarf005** [comment]
# V2

- https://kysely.dev
- https://deno.land/
